QNX support: can_be_nullptr causes build failure with packages using create_service()

can_be_nullptr causes build failure with packages using create_service()

After commit ros2/rclcpp@0750dc4, Rolling QNX build https://gitlab.com/qnx/ros2/ros2_qnx/-/jobs/1443549330 was broken due to the following error in examples_rclcpp_minimal_services:

```
Starting >>> examples_rclcpp_minimal_service
--- stderr: examples_rclcpp_minimal_service                             
In file included from /home/builder/ros2_rolling/install/aarch64le/include/rclcpp/service.hpp:28,
                 from /home/builder/ros2_rolling/install/aarch64le/include/rclcpp/callback_group.hpp:25,
                 from /home/builder/ros2_rolling/install/aarch64le/include/rclcpp/any_executable.hpp:20,
                 from /home/builder/ros2_rolling/install/aarch64le/include/rclcpp/memory_strategy.hpp:25,
                 from /home/builder/ros2_rolling/install/aarch64le/include/rclcpp/memory_strategies.hpp:18,
                 from /home/builder/ros2_rolling/install/aarch64le/include/rclcpp/executor_options.hpp:20,
                 from /home/builder/ros2_rolling/install/aarch64le/include/rclcpp/executor.hpp:37,
                 from /home/builder/ros2_rolling/install/aarch64le/include/rclcpp/executors/multi_threaded_executor.hpp:25,
                 from /home/builder/ros2_rolling/install/aarch64le/include/rclcpp/executors.hpp:21,
                 from /home/builder/ros2_rolling/install/aarch64le/include/rclcpp/rclcpp.hpp:155,
                 from /home/builder/ros2_rolling/src/ros2/examples/rclcpp/services/minimal_service/main.cpp:19:
/home/builder/ros2_rolling/install/aarch64le/include/rclcpp/any_service_callback.hpp: In instantiation of 'struct rclcpp::detail::can_be_nullptr<void (&)(std::__1::shared_ptr<rmw_request_id_t>, std::__1::shared_ptr<example_interfaces::srv::AddTwoInts_Request_<std::__1::allocator<void> > >, std::__1::shared_ptr<example_interfaces::srv::AddTwoInts_Response_<std::__1::allocator<void> > >), void>':
/home/builder/ros2_rolling/install/aarch64le/include/rclcpp/any_service_callback.hpp:104:80:   required by substitution of 'template<class CallbackT, typename std::__1::enable_if<rclcpp::detail::can_be_nullptr<CallbackT, void>::value, int>::type <anonymous> > void rclcpp::AnyServiceCallback<example_interfaces::srv::AddTwoInts>::set<CallbackT, <enumerator> >(CallbackT&&) [with CallbackT = void (&)(std::__1::shared_ptr<rmw_request_id_t>, std::__1::shared_ptr<example_interfaces::srv::AddTwoInts_Request_<std::__1::allocator<void> > >, std::__1::shared_ptr<example_interfaces::srv::AddTwoInts_Response_<std::__1::allocator<void> > >); typename std::__1::enable_if<rclcpp::detail::can_be_nullptr<CallbackT, void>::value, int>::type <anonymous> = <missing>]'
/home/builder/ros2_rolling/install/aarch64le/include/rclcpp/create_service.hpp:43:3:   required from 'typename rclcpp::Service<ServiceT>::SharedPtr rclcpp::create_service(std::__1::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>, std::__1::shared_ptr<rclcpp::node_interfaces::NodeServicesInterface>, const string&, CallbackT&&, const rmw_qos_profile_t&, rclcpp::CallbackGroup::SharedPtr) [with ServiceT = example_interfaces::srv::AddTwoInts; CallbackT = void (&)(std::__1::shared_ptr<rmw_request_id_t>, std::__1::shared_ptr<example_interfaces::srv::AddTwoInts_Request_<std::__1::allocator<void> > >, std::__1::shared_ptr<example_interfaces::srv::AddTwoInts_Response_<std::__1::allocator<void> > >); typename rclcpp::Service<ServiceT>::SharedPtr = std::__1::shared_ptr<rclcpp::Service<example_interfaces::srv::AddTwoInts> >; std::__1::string = std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >; rmw_qos_profile_t = rmw_qos_profile_t; rclcpp::CallbackGroup::SharedPtr = std::__1::shared_ptr<rclcpp::CallbackGroup>]'
/home/builder/ros2_rolling/install/aarch64le/include/rclcpp/node_impl.hpp:147:53:   required from 'typename rclcpp::Service<ServiceT>::SharedPtr rclcpp::Node::create_service(const string&, CallbackT&&, const rmw_qos_profile_t&, rclcpp::CallbackGroup::SharedPtr) [with ServiceT = example_interfaces::srv::AddTwoInts; CallbackT = void (&)(std::__1::shared_ptr<rmw_request_id_t>, std::__1::shared_ptr<example_interfaces::srv::AddTwoInts_Request_<std::__1::allocator<void> > >, std::__1::shared_ptr<example_interfaces::srv::AddTwoInts_Response_<std::__1::allocator<void> > >); typename rclcpp::Service<ServiceT>::SharedPtr = std::__1::shared_ptr<rclcpp::Service<example_interfaces::srv::AddTwoInts> >; std::__1::string = std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >; rmw_qos_profile_t = rmw_qos_profile_t; rclcpp::CallbackGroup::SharedPtr = std::__1::shared_ptr<rclcpp::CallbackGroup>]'
/home/builder/ros2_rolling/src/ros2/examples/rclcpp/services/minimal_service/main.cpp:40:82:   required from here
/home/builder/ros2_rolling/install/aarch64le/include/rclcpp/any_service_callback.hpp:44:74: error: assignment of read-only location 'std::__1::declval<void (&)(std::__1::shared_ptr<rmw_request_id_t>, std::__1::shared_ptr<example_interfaces::srv::AddTwoInts_Request_<std::__1::allocator<void> > >, std::__1::shared_ptr<example_interfaces::srv::AddTwoInts_Response_<std::__1::allocator<void> > >)>()'
     decltype(std::declval<T>() == nullptr), decltype(std::declval<T &>() = nullptr)>>
                                                      ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
/home/builder/ros2_rolling/install/aarch64le/include/rclcpp/any_service_callback.hpp:44:74: error: assignment of read-only location 'std::__1::declval<void (&)(std::__1::shared_ptr<rmw_request_id_t>, std::__1::shared_ptr<example_interfaces::srv::AddTwoInts_Request_<std::__1::allocator<void> > >, std::__1::shared_ptr<example_interfaces::srv::AddTwoInts_Response_<std::__1::allocator<void> > >)>()'
```
It seems that the compiler is refusing to accept the assignment ```decltype(std::declval<T &>() = nullptr```

QNX uses g++ v8.3.0 with LLVM libc++

After removing ```decltype(std::declval<T &>() = nullptr``` and using can_be_nullptr as below, building examples_rclcpp_minimal_service succeeds.
 
```
template<typename T>
struct can_be_nullptr<T, std::void_t<
    decltype(std::declval<T>() == nullptr)>>
  : std::true_type {};
}
```
Is it okay for QNX implementation to test only if the callback function is equal to nullptr?

Hello,
I am wondering whats the status of this, I am getting the same issue trying to build rclcpp/rclcpp/test/benchmark:
```
In file included from /home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/service.hpp:28,
                 from /home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/callback_group.hpp:25,
                 from /home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/any_executable.hpp:20,
                 from /home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/memory_strategy.hpp:25,
                 from /home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/memory_strategies.hpp:18,
                 from /home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/executor_options.hpp:20,
                 from /home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/executor.hpp:37,
                 from /home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp:25,
                 from /home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/executors.hpp:21,
                 from /home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/rclcpp.hpp:155,
                 from /home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/test/benchmark/benchmark_client.cpp:19:
/home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/any_service_callback.hpp: In instantiation of ‘struct rclcpp::detail::can_be_nullptr<ClientPerformanceTest::SetUp(benchmark::State&)::<lambda(test_msgs::srv::Empty_Request_<std::allocator<void> >::SharedPtr, test_msgs::srv::Empty_Response_<std::allocator<void> >::SharedPtr)>&, void>’:
/home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/any_service_callback.hpp:104:80:   required by substitution of ‘template<class CallbackT, typename std::enable_if<rclcpp::detail::can_be_nullptr<CallbackT, void>::value, int>::type <anonymous> > void rclcpp::AnyServiceCallback<test_msgs::srv::Empty>::set<CallbackT, <enumerator> >(CallbackT&&) [with CallbackT = ClientPerformanceTest::SetUp(benchmark::State&)::<lambda(test_msgs::srv::Empty_Request_<std::allocator<void> >::SharedPtr, test_msgs::srv::Empty_Response_<std::allocator<void> >::SharedPtr)>&; typename std::enable_if<rclcpp::detail::can_be_nullptr<CallbackT, void>::value, int>::type <anonymous> = <missing>]’
/home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/create_service.hpp:43:3:   required from ‘typename rclcpp::Service<ServiceT>::SharedPtr rclcpp::create_service(std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>, std::shared_ptr<rclcpp::node_interfaces::NodeServicesInterface>, const string&, CallbackT&&, const rmw_qos_profile_t&, rclcpp::CallbackGroup::SharedPtr) [with ServiceT = test_msgs::srv::Empty; CallbackT = ClientPerformanceTest::SetUp(benchmark::State&)::<lambda(test_msgs::srv::Empty_Request_<std::allocator<void> >::SharedPtr, test_msgs::srv::Empty_Response_<std::allocator<void> >::SharedPtr)>&; typename rclcpp::Service<ServiceT>::SharedPtr = std::shared_ptr<rclcpp::Service<test_msgs::srv::Empty> >; std::__cxx11::string = std::__cxx11::basic_string<char>; rmw_qos_profile_t = rmw_qos_profile_t; rclcpp::CallbackGroup::SharedPtr = std::shared_ptr<rclcpp::CallbackGroup>]’
/home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/node_impl.hpp:147:53:   required from ‘typename rclcpp::Service<ServiceT>::SharedPtr rclcpp::Node::create_service(const string&, CallbackT&&, const rmw_qos_profile_t&, rclcpp::CallbackGroup::SharedPtr) [with ServiceT = test_msgs::srv::Empty; CallbackT = ClientPerformanceTest::SetUp(benchmark::State&)::<lambda(test_msgs::srv::Empty_Request_<std::allocator<void> >::SharedPtr, test_msgs::srv::Empty_Response_<std::allocator<void> >::SharedPtr)>&; typename rclcpp::Service<ServiceT>::SharedPtr = std::shared_ptr<rclcpp::Service<test_msgs::srv::Empty> >; std::__cxx11::string = std::__cxx11::basic_string<char>; rmw_qos_profile_t = rmw_qos_profile_t; rclcpp::CallbackGroup::SharedPtr = std::shared_ptr<rclcpp::CallbackGroup>]’
/home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/test/benchmark/benchmark_client.cpp:40:93:   required from here
/home/localstorage/workspace/ros2/rolling/distro/src/ros2/rclcpp/rclcpp/include/rclcpp/any_service_callback.hpp:44:74: error: no match for ‘operator=’ (operand types are ‘ClientPerformanceTest::SetUp(benchmark::State&)::<lambda(test_msgs::srv::Empty_Request_<std::allocator<void> >::SharedPtr, test_msgs::srv::Empty_Response_<std::allocator<void> >::SharedPtr)>’ and ‘std::nullptr_t’)
     decltype(std::declval<T>() == nullptr), decltype(std::declval<T &>() = nullptr)>>
```


Is the fix mentioned by  @asobhy-qnx  safe, does it break anything?

--- a/rclcpp/include/rclcpp/any_service_callback.hpp
+++ b/rclcpp/include/rclcpp/any_service_callback.hpp
@@ -40,9 +40,14 @@ struct can_be_nullptr : std::false_type {};
 // but we see a warning that they can never be null when using it.
 // We also test if `T &` can be assigned to `nullptr` to avoid the issue.
 template<typename T>
+#ifdef __QNXNTO__
+struct can_be_nullptr<T, std::void_t<
+    decltype(std::declval<T>() == nullptr)>>: std::true_type {};
+#else
 struct can_be_nullptr<T, std::void_t<
     decltype(std::declval<T>() == nullptr), decltype(std::declval<T &>() = nullptr)>>
   : std::true_type {};
+#endif
 }  // namespace detail
 
 // Forward declare

